File: pytest/test_i2c_ADC_6axis.py
import sys
import os
import time
from datetime import datetime
import numpy as np
sys.path.append("../build/")

import debugger

def generate_filename(rootpath):
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    filename = os.path.join(rootpath, "voltages_{}.txt".format(timestamp))
    return filename

if __name__ == "__main__":
    sensor = debugger.Sensor_ADC()
    sensor.open_sensor(0)

    os.makedirs("../temp", exist_ok=True)
    file_path = generate_filename("../temp")
    
    # 打开文件以写入
    try:
        with open(file_path, "w") as outFile:
            for i in range(100):
                voltages = []
                for device in range(4):
                    for channel in range(2):
                        voltage = sensor.get_voltage_mux(device, channel)
                        voltages.append(voltage)

                for voltage in voltages:
                    print("{:5.2f}  ".format(voltage), end="")
                    outFile.write("{} ".format(voltage))
                print()
                outFile.write("\n")

                time.sleep(0.005)  # 5 毫秒的延迟

    except IOError:
        print("无法打开文件！")

File: test/devices/test_i2c_ADC_6axis.cpp
#include "mraa.hpp"
#include <iostream>
#include <fstream>
#include <unistd.h>
#include <vector>
#include "lib_io.h"
#include "sensor_adc.h"

int main() {
    std::ofstream outFile("../temp/voltages.txt");
    if (!outFile) {
        std::cerr << "cannot open file!" << std::endl;
    }

    Sensor_ADC sensor;
    sensor.open_sensor(0);
    while (1) {
        std::vector<float> voltages;
        for (int device = 0; device < 4; device++) {
            for (int channel = 0; channel < 2; channel++) {
                float voltage = sensor.get_voltage_mux(device, channel);
                voltages.emplace_back(voltage);
            }
        }
        for (int i = 0; i < voltages.size(); i++) {
            printf("%5.2f  ", voltages[i]);
            outFile << voltages[i] << " ";
        }
        printf("\n");
        outFile << std::endl;

        std::this_thread::sleep_for(std::chrono::milliseconds(5));
    }
    outFile.close();
    return 0;
}